src/safe_mpc/controller.py

```diff
@@ -0,0 +1,121 @@
+import numpy as np
+from .abstract import AbstractController
+
+
+class NaiveController(AbstractController):
+    def __init__(self, simulator):
+        super().__init__(simulator)
+
+    def checkGuess(self, x, u):
+        return self.model.checkRunningConstraints(x, u)
+
+    def initialize(self, x0):
+        # Trivial guess
+        self.x_guess = np.full((self.N + 1, self.model.nx), x0)
+        self.u_guess = np.zeros((self.N, self.model.nu))
+        # Solve the OCP
+        status = self.solve(x0)
+        # self.ocp_solver.print_statistics()
+        # print('status = ', status)
+        if status == 0 or status == 2:
+            for i in range(self.N):
+                self.x_guess[i] = self.ocp_solver.get(i, "x")
+                self.u_guess[i] = self.ocp_solver.get(i, "u")
+            self.x_guess[self.N] = self.ocp_solver.get(self.N, "x")
+
+            if self.checkGuess(self.x_guess, self.u_guess): #and \
+               # self.simulator.checkDynamicsConstraints(self.x_guess, self.u_guess):
+                self.success += 1
+            else:
+                self.x_guess *= np.nan
+                self.u_guess *= np.nan
+
+    def step(self, x):
+        status = self.solve(x)
+        u = self.ocp_solver.get(0, "u")
+        if status == 0 and self.model.checkControlConstraints(u):
+            self.fails = 0
+        else:
+            if self.fails >= self.N:
+                return None
+            self.fails += 1
+        return self.provideControl()
+
+
+class STController(NaiveController):
+    def __init__(self, simulator):
+        super().__init__(simulator)
+
+    def additionalSetting(self):
+        self.terminalConstraint()
+
+
+class STWAController(STController):
+    def __init__(self, simulator):
+        super().__init__(simulator)
+        self.x_viable = None                # TODO: to be initialized
+
+    def step(self, x):
+        status = self.solve(x)
+        if status == 0 and self.model.checkRunningConstraints(self.x_temp, self.u_temp) and \
```

Comment:
self.x_temp should contain the whole state trajectory, but the method checkRunningConstraints seems to be checking a single state-control pair. How can this work?

Reply:
`checkRunningConstraints` checks on all the trajectory, for a single state-control pair we need to provide only that pair as input of the method